hugo/content/3-queries/introduction.md

```diff
@@ -0,0 +1,169 @@
+
+#### Referring to embedded fields
+
+Referring to fields in embedded objects is a bit different.
+Let's say we wanted to add an Address class to our User:
+
+```cpp
+class Address {
+public:
+    std::string street;
+    std::string city;
+    std::string state;
+    std::string zip;  
+
+    MANGROVE_MAKE_KEYS(Address, MANGROVE_NVP(street), MANGROVE_NVP(city),
+                                MANGROVE_NVP(state), MANGROVE_NVP(zip));
+};
+
+class User {
+public:
+    std::string username;
+    std::int32_t age;
+    Address addr;
+
+    MANGROVE_MAKE_KEYS(User, MANGROVE_NVP(username), MANGROVE_NVP(age), MANGROVE_NVP(addr));
+}
+```
+
+Now, we can query for users using their address using the `MANGROVE_CHILD` macro:
+
+```cpp
+// { "addr.zip": "21211" }
+auto results = collection.find(MANGROVE_CHILD(User, addr, zip) == "21211");
+```
+The `MANGROVE_CHILD` takes a base class, and then the names of each successive nested field.
+
+{{% notice info %}}
+Mangrove can support as many levels of nesting as is allowed in the {{% a_blank "BSON specification" "https://docs.mongodb.com/manual/reference/limits/#Nested-Depth-for-BSON-Documents" %}}.
+{{% /notice %}}
+
+Another way to refer to children is to link name-value pairs using the `->*` operator:
+
+```cpp
+// { "addr.zip": "21211" }
+auto results = collection.find(MANGROVE_KEY(User::addr) ->* MANGROVE_KEY(Address::zip) == "21211");
+```
+
+#### Referring to array elements
+
+Referring to specific array elements in queries is the same as in any C++ expression ---
+one can simply user the square brackets `[]` to specify a certain index.
+For instance, the following query finds users whose 10th score in some game is higher than 1000:
+
+```cpp
+// { "scores.9": {$gt: 1000} }
+// (note that arrays are 0-indexed)
+auto results = collection.find(MANGROVE_KEY(User::scores)[9] > 1000);
+```
+
+## Type checking
+
+Mangrove provides type checking on queries, so that comparing fields and values with mismatched types,
+or using array operators on non-array fields, results in syntax errors.
+
+For instance, the following query will fail at compile time:
+
+```cpp
+// { age: {$gte: "21"} }
+auto results = collection.find(MANGROVE_KEY(User::age) >= "21");
+```
+
+In the mongo shell, for instance, such a query would be just fine,
+but would have done a lexicographic comparison between the `age` field and the string `"21"`.
+This is definitely not what we want, and may only have been discovered later, in production.
+
+Similarly, the query
+
+```cpp
+// { age: {$all: [15, 17, 19]} }
+auto results = collection.find(MANGROVE_KEY(User::age).all(std::vector<int>{15, 17, 19}));
+```
+
+does not make sense because `User::age` is not an array, and will cause a syntax error accordingly.
+
+## Combining queries
+
+There are several ways to combine different conditions into one query in Mangrove.
+The simplest one is the comma operator --- conditions can be appended one after the other, separated by a comma.
```

Comment:
The comma is the same as appending fields in a BSON query, i.e. `{ age: 21, state: NY }`. I'll clarify this here.